File: core/src/main/java/ru/kuchanov/scpcore/api/model/remoteconfig/LevelsJson.java
package ru.kuchanov.scpcore.api.model.remoteconfig;

import android.support.annotation.Nullable;

import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.Collections;
import java.util.List;

import ru.kuchanov.scpcore.Constants;
import timber.log.Timber;

/**
 * Created by mohax on 29.04.2017.
 * <p>
 * for scp_ru
 */
public class LevelsJson {

    public static final int MAX_LEVEL_ID = 5;
    public static final int NO_SCORE_TO_MAX_LEVEL = -1;

    public List<Level> levels;

    /**
     * returns NO_SCORE_TO_MAX_LEVEL if level is already MAX_LEVEL_ID
     */
    public int scoreToNextLevel(int userScore, Level curLevel) {
        if (curLevel.id == MAX_LEVEL_ID) {
            return NO_SCORE_TO_MAX_LEVEL;
        }
        LevelsJson.Level nextLevel = levels.get(curLevel.id + 1);

        int nextLevelScore = nextLevel.score;

        int max = nextLevelScore - curLevel.score;
        int value = userScore - curLevel.score;

        return max - value;
    }

    public int getLevelMaxScore(Level curLevel) {
        if (curLevel.id == MAX_LEVEL_ID) {
            return NO_SCORE_TO_MAX_LEVEL;
        }
        LevelsJson.Level nextLevel = levels.get(curLevel.id + 1);

        int nextLevelScore = nextLevel.score;

        return nextLevelScore - curLevel.score;
    }

    @Nullable
    public Level getLevelForScore(int score) {
        Level userLevel = null;
        String levelsJsonString = FirebaseRemoteConfig.getInstance().getString(Constants.Firebase.RemoteConfigKeys.LEVELS_JSON);
        LevelsJson levelsJson = new GsonBuilder().create().fromJson(levelsJsonString, LevelsJson.class);
        if (levelsJson != null) {
            Collections.reverse(levelsJson.levels);
            for (int i = 0; i < levelsJson.levels.size(); i++) {
                LevelsJson.Level level = levelsJson.levels.get(i);
                if (score >= level.score) {
                    userLevel = level;
                    break;
                } else if (i == levelsJson.levels.size() - 1) {
                    //so max level reached
                    userLevel = level;
                }
            }
        } else {
            Timber.e("levelsJson is null");
        }
        return userLevel;
    }

    public static LevelsJson getLevelsJson() {
        String levelsJsonString = FirebaseRemoteConfig.getInstance().getString(Constants.Firebase.RemoteConfigKeys.LEVELS_JSON);
        return new Gson().fromJson(levelsJsonString, LevelsJson.class);
    }

    public static class Level {

        public int id;
        public String title;
        public int score;

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Level level = (Level) o;

            return id == level.id;
        }

        @Override
        public int hashCode() {
            return id;
        }

        @Override
        public String toString() {
            return "Level{" +
                    "id=" + id +
                    ", title='" + title + '\'' +
                    ", score=" + score +
                    '}';
        }
    }

    @Override
    public String toString() {
        return "LevelsJson{" +
                "levels=" + levels +
                '}';
    }
}

File: core/src/main/java/ru/kuchanov/scpcore/api/model/remoteconfig/LevelsJson.kt
package ru.kuchanov.scpcore.api.model.remoteconfig

import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.gson.Gson
import ru.kuchanov.scpcore.Constants

/**
 * Created by mohax on 29.04.2017.
 *
 *
 * for scp_ru
 */
data class LevelsJson(val levels: List<Level>) {

    /**
     * returns NO_SCORE_TO_MAX_LEVEL if level is already MAX_LEVEL_ID
     */
    fun scoreToNextLevel(userScore: Int, curLevel: Level): Int {
        if (curLevel.id == MAX_LEVEL_ID) {
            return NO_SCORE_TO_MAX_LEVEL
        }
        val nextLevel = levels[curLevel.id + 1]

        val nextLevelScore = nextLevel.score

        val max = nextLevelScore - curLevel.score
        val value = userScore - curLevel.score

        return max - value
    }

    fun getLevelMaxScore(curLevel: Level): Int {
        if (curLevel.id == MAX_LEVEL_ID) {
            return NO_SCORE_TO_MAX_LEVEL
        }
        val nextLevel = levels[curLevel.id + 1]

        val nextLevelScore = nextLevel.score

        return nextLevelScore - curLevel.score
    }

    fun getLevelForScore(score: Int): Level? = levels.findLast { score > it.score }

    data class Level(val id: Int,
        val title: String,
        val score: Int
    )

    companion object {

        @Suppress("MayBeConstant")
        @JvmField
        val MAX_LEVEL_ID = 5
        const val NO_SCORE_TO_MAX_LEVEL = -1

        private const val LEVELS_JSON_DEFAULT = """{"levels":[{"id":0,"title":"Level 0 (For Official Use Only)","score":0},{"id":1,"title":"Level 1 (Confidential)","score":1000},{"id":2,"title":"Level 2 (Restricted)","score":2000},{"id":3,"title":"Level 3 (Secret)","score":3000},{"id":4,"title":"Level 4 (Top Secret)","score":6000},{"id":5,"title":"Level 5 (Thaumiel)","score":10000}]}"""

        @JvmStatic
        val levelsJson: LevelsJson
            get() {
                val levelsJsonString = FirebaseRemoteConfig.getInstance().getString(Constants.Firebase.RemoteConfigKeys.LEVELS_JSON) ?: LEVELS_JSON_DEFAULT
                return Gson().fromJson(levelsJsonString, LevelsJson::class.java)
            }
    }
}